flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/RetryingExecutor.java

```diff
@@ -170,16 +189,24 @@ private RetriableTask(
 
         @Override
         public void run() {
-            LOG.debug("starting attempt {}", current);
+            LOG.debug("starting attempt {}", attemptNumber);
             if (!actionCompleted.get()) {
                 Optional<ScheduledFuture<?>> timeoutFuture = scheduleTimeout();
                 try {
-                    runnable.run();
+                    Result result = action.tryExecute();
                     if (actionCompleted.compareAndSet(false, true)) {
-                        LOG.debug("succeeded with {} attempts", current);
-                        attemptsPerTaskHistogram.update(current);
+                        LOG.debug("succeeded with {} attempts", attemptNumber);
+                        action.completeWithResult(result);
+                        attemptsPerTaskHistogram.update(attemptNumber);
+                    } else {
+                        LOG.debug(
+                                "discard unnecessarily uploaded state, attempt {}", attemptNumber);
+                        try {
+                            action.discardResult(result);
+                        } catch (Exception e) {
+                            LOG.warn("unable to discard execution attempt result", e);
+                        }
                     }
-                    attemptCompleted.set(true);
                 } catch (Exception e) {
                     handleError(e);
```

Comment:
Just a question, how will we handle error if we could `actionCompleted.compareAndSet(false, true)` but throwing a unchecked exception within the suceess statement.